### PR TITLE
ARROW-13055: [Doc] Create canonical extension types document

### PR DESCRIPTION
**Pull request by pitrou**
Vote result at https://lists.apache.org/thread/sxd5fhc42hb6svs79t3fd79gkqj83pfh

### CODE REVIEW

**Comment by pitrou** [comment]
@wjones127 Would you like to review this?

File: docs/source/format/CanonicalExtensions.rst [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+Procedures
+==========
+
+These rules must be followed for the standardization of canonical extension
+types:
+
+* Canonical extension types are described and maintained below in this document.
+
+* Each canonical extension type requires a distinct discussion and vote
+  on the `Arrow development mailing-list <https://arrow.apache.org/community/>`__.
```
Comment (wjones127) [inline docs/source/format/CanonicalExtensions.rst:44]: typo?
```suggestion
  on the `Arrow development mailing-list <https://arrow.apache.org/community/>`_.
```

File: docs/source/format/CanonicalExtensions.rst [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+* The specification text to be added *must* follow these requirements:
+
+  1) It *must* define a well-defined extension name starting with "``arrow.``".
+
+  2) Its parameters, if any, *must* be described in the proposal.
+
+  3) Its serialization *must* be described in the proposal and should
+     not require unduly work or unusual software dependencies (for example, a
```
Comment (wjones127) [inline docs/source/format/CanonicalExtensions.rst:53]: What do you mean by "unduly work"? Is that in terms of readability to reviewers? Or in terms of computation time?

**Review by lidavidm — APPROVED** [review]
LGTM. 

A few things:
- Weston had some comments in the vote thread, will those be integrated?

  > This is maybe implied but I would add that
modification of extension types must also require a vote and should be
backwards compatible.  Furthermore, extension types (particularly
those with extensive parameterization/serialization should discuss how
future additions would be made.  For example, if serialized via JSON
than readers should tolerate (and ignore) unexpected keys.

- Maybe add 'canonical extension type' to Glossary.rst as well?

**Review by wjones127 — COMMENTED** [review]
This looks pretty good. Once we have our first canonical type, hopefully we will have a model proposal on the ML we can link to. It would be nice, for example, to show what kind of detail the semantics should be described in.

File: docs/source/format/CanonicalExtensions.rst (hunk 2 above)
Comment (pitrou) [reply docs/source/format/CanonicalExtensions.rst:53]: I meant in terms of implementation effort.

Comment (pitrou) [reply docs/source/format/CanonicalExtensions.rst:53]: I can make this more explicit, though ideally we should not deviate too much from what has been voted on on the ML.

File: docs/source/format/CanonicalExtensions.rst (hunk 1 above)
Comment (pitrou) [reply docs/source/format/CanonicalExtensions.rst:44]: Not really a typo, with a double trailing underscore it's an "anonymous" link target, but this doesn't really matter and I can remove one underscore for readability :-)

Comment (wjones127) [reply docs/source/format/CanonicalExtensions.rst:44]: Oh that's fine as is. Hadn't seen that before.

**Comment by pitrou** [comment]
@lidavidm @wjones127 I believe I addressed your comments, could you take another look?

**Review by lidavidm — APPROVED** [review]

**Review by jorisvandenbossche — APPROVED** [review]

**Pull request merged**